src/components/macro/macro-action-editor/macro-action-editor.component.ts

```diff
@@ -0,0 +1,100 @@
+import { Component, EventEmitter, Input, OnInit, Output, ViewChild } from '@angular/core';
+
+import { MacroAction, TextMacroAction, macroActionType } from '../../../config-serializer/config-items/macro-action';
+import { EditableMacroAction } from '../../../config-serializer/config-items/macro-action/EditableMacroAction';
+import { MacroKeyTabComponent } from './tab/macro-key';
+
+enum TabName {
+    Keypress,
+    Text,
+    Mouse,
+    Delay
+};
+
+@Component({
+    selector: 'macro-action-editor',
+    template: require('./macro-action-editor.component.html'),
+    styles: [require('./macro-action-editor.component.scss')],
+    host: { 'class': 'macro-action-editor' }
+})
+export class MacroActionEditorComponent implements OnInit {
+    @Input() macroAction: MacroAction;
+
+    @Output() save = new EventEmitter<MacroAction>();
+    @Output() cancel = new EventEmitter<void>();
+
+    @ViewChild('tab') selectedTab: any;
+
+    private editableMacroAction: EditableMacroAction;
+    private activeTab: TabName;
+    /* tslint:disable:variable-name: It is an enum type. So it can start with uppercase. */
+    /* tslint:disable:no-unused-variable: It is used in the template. */
+    private TabName = TabName;
+    /* tslint:enable:no-unused-variable */
+    /* tslint:enable:variable-name */
+
+    ngOnInit() {
+        let macroAction: MacroAction = this.macroAction ? this.macroAction : new TextMacroAction();
+        this.editableMacroAction = new EditableMacroAction(macroAction.toJsObject());
+        let tab: TabName = this.getTabName(this.editableMacroAction);
+        this.activeTab = tab;
+    }
+
+    onCancelClick(): void {
+        this.cancel.emit();
+    }
+
+    onSaveClick(): void {
+        try {
+            const action = this.editableMacroAction;
+            if (action.isKeyAction()) {
+                // Could updating the saved keys be done in a better way?
+                const tab = this.selectedTab as MacroKeyTabComponent;
```

Comment:
It should work fine because of the isKeyAction check as changing the tab means also changing EditableMacroAction.macroActionType. So it should not be possible to select a different tab and have isKeyAction pass.

The real solution would be to update the EditableMacroAction every time something is changed in the key press component by providing an output event that gets run every time you change something in it. The reason I didn't do this is because I wanted to keep my changes in the macro components as much as possible and not mess with the other stuff.